### PR TITLE
[WIP] fix comment printing in decorators

### LINKED ISSUE #1460: Comments inside decorator arguments are moved to the wrong line
Babylon seems to support decorators, but prettier incorrectly relocates comments inside of decorator arguments. Please see [this example](https://prettier.github.io/prettier/#%7B%22content%22%3A%22var%20x%20%3D%20100%5Cn%5Cn%40Hello(%7B%5Cn%20%20a%3A%20'a'%2C%20%2F%2F%20Comment%20is%20in%20the%20wrong%20place%5Cn%20%20%2F%2F%20test%5Cn%20%20b%3A%20'2'%5Cn%7D)%5Cnclass%20X%20%7B%5Cn%20%20%5Cn%7D%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22es5%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22semi%22%3Afalse%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%7D%7D). Notice that the comments are moved to the wrong lines.

### Input
```javascript
var x = 100

@Hello({
  a: 'a', // Comment is in the wrong place
  // test
  b: '2'
})
class X {
  
}
```

### Output

```javascript
var x = 100 // Comment is in the wrong place

// test
@Hello({
  a: "a",
  b: "2",
})
class X {}
```

```shell
$ prettier --version
1.2.2
```

### PR DESCRIPTION
**Pull request by despairblue**
fix #1460 

This still needs some work. The hack in `getSortedChildNodes` needs to be fixed properly, probably in `ast-types` or via forking.

This also still fails for things like: 

```ts
@AutoSubscribeStore
//store in resub, is a class extends from StoreBase.
class HelloStore extends StoreBase {
    getHello():string {
        return this._helloString;
    }
}
```
:arrow_down: 
```ts
@AutoSubscribeStore
@SomeOtherMagic
//store in resub, is a class extends from StoreBase.
// apply differnet magic
class HelloStore extends StoreBase {
  getHello(): string {
    return this._helloString;
  }
}
```

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -98,8 +98,7 @@ function genericPrint(path, options, printPath, args) {
 
       if (
         node.decorators.length === 1 &&
-        (decorator.type === "Identifier" ||
```
Comment (vjeux) [inline src/printer.js]: This is a breaking change. Aren't single identifier decorators supposed to be inlined in mobx?

File: src/comments.js [hunk 2]
```diff
@@ -161,6 +165,7 @@ function attach(comments, ast, text) {
           comment
         ) ||
         handleTryStatementComments(enclosingNode, followingNode, comment) ||
+        handleDecoratorComments(enclosingNode, comment) ||
```
Comment (vjeux) [inline src/comments.js]: what happens if you don't put this?

File: src/printer.js (hunk 1 above)
Comment (despairblue) [inline src/printer.js]: I wasn't sure if this was on purpose. It just looked weird to me that the class keyword wasn't the first on the line. I'm not using them, it if that's how they are formatted I put it back in :)

File: src/comments.js (hunk 2 above)
Comment (despairblue) [inline src/comments.js]: The comment moves above all decorators.

**Comment by vjeux** [comment]
I think another issue is that the node size doesn't take into account comments. So if you do a binary search, you won't find the right one to go into.

File: src/printer.js (hunk 1 above)
Comment (despairblue) [inline src/printer.js]: I put it back in but made an exception for classes:

```ts
@Decorarte
class Foo {}
```

There will always be a hardline after a decorator if it decorates a class. That behaviour is consistent with how exported class declarations are handled.

File: src/comments.js (hunk 2 above)
Comment (despairblue) [inline src/comments.js]:
```ts
@AutoSubscribeStore
//store in resub, is a class extends from StoreBase.
class HelloStore extends StoreBase {
    @autoSubscribe
    // this method should return a string type
    getHello():string {
        return this._helloString;
    }
}
```
:arrow_down:
```ts
//store in resub, is a class extends from StoreBase.
@AutoSubscribeStore
class HelloStore extends StoreBase {
  @autoSubscribe // this method should return a string type
  getHello(): string {
    return this._helloString;
  }
}
```

But I see that it does not really handle the decorated methods well when I don't add a newline after the decorators.

I'll keep working on this, but I can only invest a couple of minutes per day since I'm actually on a vacation :smile:

File: src/printer.js [hunk 3]
```diff
@@ -97,6 +97,7 @@ function genericPrint(path, options, printPath, args) {
       }
 
       if (
+        node.type !== "ClassDeclaration" &&
```
Comment (vjeux) [inline src/printer.js]: This was done in another PR

File: src/comments.js [hunk 4]
```diff
@@ -52,6 +52,10 @@ function getSortedChildNodes(node, text, resultArray) {
     names = Object.keys(node);
   } else if (isObject.check(node)) {
     names = types.getFieldNames(node);
+
+    if (node.decorators) {
+      names.push("decorators");
+    }
```
Comment (vjeux) [inline src/comments.js]: This is no longer an issue now that we moved off of ast-types

**Comment by vjeux** [comment]
I'm going to close this as 2/3 of it were fixed. If you want to open another PR to fix the remaining piece, that would be awesome. Even though this didn't land, this was a great inspiration to get decorator comments fixed for the 1.4 release :)
